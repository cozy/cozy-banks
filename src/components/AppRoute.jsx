import React from 'react'
import { Navigate, Outlet, Route, Routes } from 'react-router-dom'
import App from 'components/App'
import { isWebApp } from 'cozy-device-helper'

import { CategoriesPage } from 'ducks/categories'
import {
  Settings,
  AccountsSettings,
  GroupsSettings,
  ExistingGroupSettings,
  NewGroupSettings,
  TagsSettings,
  Configuration
} from 'ducks/settings'
import { Balance, BalanceDetailsPage } from 'ducks/balance'
import {
  DebugRecurrencePage,
  RecurrencesPage,
  RecurrencePage
} from 'ducks/recurrence'
import { TransferPage } from 'ducks/transfers'
import { SearchPage } from 'ducks/search'
import { AnalysisPage } from 'ducks/analysis'
import UserActionRequired from 'components/UserActionRequired'
import ScrollToTopOnMountWrapper from 'components/scrollToTopOnMount'
import PlannedTransactionsPage from 'ducks/future/PlannedTransactionsPage'
import SetFilterAndRedirect from 'ducks/balance/SetFilterAndRedirect'
import TagPage from 'ducks/tags/TagPage'
import Export from 'ducks/settings/Export'
import Import from 'ducks/settings/Import/Import'

const OutletWrapper = ({ Component }) => (
  <>
    <Component />
    <Outlet />
  </>
)

// Use a function to delay instantation and have access to AppRoute methods
const AppRoute = () => {
  const { routes, condition } = AppRoute.renderExtraRoutesOnly()
  if (condition) {
    return (
      <Routes>
        <Route element={<UserActionRequired />}>{routes}</Route>
      </Routes>
    )
  }

  return (
    <Routes>
      <Route element={<UserActionRequired />}>
        <Route path="/" element={<App />}>
          {isWebApp() && (
            <Route index element={<Navigate to="balances" replace />} />
          )}
          <Route path="balances">
            <Route
              index
              element={
                <ScrollToTopOnMountWrapper>
                  <Balance />
                </ScrollToTopOnMountWrapper>
              }
            />
            <Route
              path="details"
              element={
                <ScrollToTopOnMountWrapper>
                  <BalanceDetailsPage />
                </ScrollToTopOnMountWrapper>
              }
            />
            <Route
              path="future"
              element={
                <ScrollToTopOnMountWrapper>
                  <PlannedTransactionsPage />
                </ScrollToTopOnMountWrapper>
              }
            />
            <Route
              path=":accountOrGroupId/:page"
              element={<SetFilterAndRedirect />}
            />
          </Route>
          <Route
            path="categories/*"
            element={<Navigate to="../analysis/categories" replace />}
          ></Route>
          <Route
            path="recurrence/*"
            element={<Navigate to="../analysis/recurrence" replace />}
          ></Route>
          <Route
            path="analysis"
            element={
              <ScrollToTopOnMountWrapper>
                <AnalysisPage />
              </ScrollToTopOnMountWrapper>
            }
          >
            <Route path="categories">
              <Route
                index
                element={
                  <ScrollToTopOnMountWrapper>
                    <CategoriesPage />
                  </ScrollToTopOnMountWrapper>
                }
              />
              <Route
                path=":categoryName"
                element={
                  <ScrollToTopOnMountWrapper>
                    <CategoriesPage />
                  </ScrollToTopOnMountWrapper>
                }
              />
              <Route
                path=":categoryName/:subcategoryName"
                element={
                  <ScrollToTopOnMountWrapper>
                    <CategoriesPage />
                  </ScrollToTopOnMountWrapper>
                }
              />
            </Route>
            <Route path="recurrence">
              <Route
                index
                element={
                  <ScrollToTopOnMountWrapper>
                    <RecurrencesPage />
                  </ScrollToTopOnMountWrapper>
                }
              />
              <Route
                path=":bundleId"
                element={
                  <ScrollToTopOnMountWrapper>
                    <RecurrencePage />
                  </ScrollToTopOnMountWrapper>
                }
              />
            </Route>
          </Route>
          <Route path="settings">
            <Route
              path="import"
              element={<Navigate to="../configuration/import" replace />}
            />
            <Route
              path="configuration/export"
              element={<Navigate to="../export" replace />}
            />
            <Route path="export" element={<Export />} />
            <Route
              element={
                <ScrollToTopOnMountWrapper>
                  <Settings />
                </ScrollToTopOnMountWrapper>
              }
            >
              <Route index element={<Configuration />} />
              <Route path="accounts" element={<AccountsSettings />} />
              <Route path="groups" element={<GroupsSettings />} />
              <Route path="tags" element={<TagsSettings />} />
              <Route
                path="configuration"
                element={<OutletWrapper Component={Configuration} />}
              >
                <Route path="import" element={<Import />} />
              </Route>
            </Route>
            <Route
              path="groups/new"
              element={
                <ScrollToTopOnMountWrapper>
                  <NewGroupSettings />
                </ScrollToTopOnMountWrapper>
              }
            />
            <Route
              path="groups/:groupId"
              element={
                <ScrollToTopOnMountWrapper>
                  <ExistingGroupSettings />
                </ScrollToTopOnMountWrapper>
              }
            />
            <Route
              path="accounts/:accountId"
              element={<Navigate to="../accounts" replace />}
            />
          </Route>
          <Route
            path="tag/:tagId"
            element={
              <ScrollToTopOnMountWrapper>
                <TagPage />
              </ScrollToTopOnMountWrapper>
            }
          />
          <Route
            path="transfers"
            element={
              <ScrollToTopOnMountWrapper>
                <TransferPage />
              </ScrollToTopOnMountWrapper>
            }
          />
          <Route
            path="search"
            element={
              <ScrollToTopOnMountWrapper>
                <SearchPage />
              </ScrollToTopOnMountWrapper>
            }
          />
          <Route
            path="search/:search"
            element={
              <ScrollToTopOnMountWrapper>
                <SearchPage />
              </ScrollToTopOnMountWrapper>
            }
          />
          <Route
            path="recurrencedebug"
            element={
              <ScrollToTopOnMountWrapper>
                <DebugRecurrencePage />
              </ScrollToTopOnMountWrapper>
            }
          />
          {AppRoute.renderExtraRoutes()}
          {isWebApp() && (
            <Route path="*" element={<Navigate to="balances" replace />} />
          )}
        </Route>
      </Route>
    </Routes>
  )
}

// Ability to overrides easily
AppRoute.renderExtraRoutes = () => null
AppRoute.renderExtraRoutesOnly = () => ({ routes: [], conditon: false })

export default AppRoute
