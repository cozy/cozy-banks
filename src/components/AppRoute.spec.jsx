import React from 'react'
import { Route } from 'react-router-dom'

import AppRoute from './AppRoute'

const NewComponent = () => <div>New component</div>

// Mock otherwise we have a fetch is not defined error due to pouchdb-browser.
// Here we are not concerned with this component
jest.mock('cozy-pouch-link', () => () => null)

// Mock deep component to avoid error when mouting it.
// Here we are not concerned with this component
jest.mock('../ducks/settings/Configuration.jsx', () => () => null)

describe('App route', () => {
  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('should be a function returning a route', () => {
    const route = AppRoute()
    expect(route).toMatchSnapshot()
  })

  it('should have renderExtraRoutes', () => {
    jest
      .spyOn(AppRoute, 'renderExtraRoutes')
      .mockReturnValue(<Route element={<NewComponent />} path="extra-route" />)
    const route = AppRoute()
    expect(route).toMatchSnapshot()
  })

  it('should have renderExtraRoutesOnly if condition is true', () => {
    jest.spyOn(AppRoute, 'renderExtraRoutesOnly').mockReturnValue({
      routes: <Route element={<NewComponent />} path="extra-route-only" />,
      condition: true
    })
    const route = AppRoute()
    expect(route).toMatchSnapshot()
  })

  it('should not have renderExtraRoutesOnly if condition is false', () => {
    jest.spyOn(AppRoute, 'renderExtraRoutesOnly').mockReturnValue({
      routes: <Route element={<NewComponent />} path="extra-route-only" />,
      condition: false
    })
    const route = AppRoute()
    expect(route).toMatchSnapshot()
  })
})
