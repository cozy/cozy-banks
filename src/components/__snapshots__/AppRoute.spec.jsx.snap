// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App route should be a function returning a route 1`] = `
<Routes>
  <Route
    element={<withClient(UserActionRequired) />}
  >
    <Route
      element={<withQuery(App) />}
      path="/"
    >
      <Route
        element={
          <Navigate
            replace={true}
            to="balances"
          />
        }
        index={true}
      />
      <Route
        path="balances"
      >
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <BalanceWithBanksJobs />
            </ScrollToTopOnMountWrapper>
          }
          index={true}
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(Connect(RawBalanceDetailsPage)) />
            </ScrollToTopOnMountWrapper>
          }
          path="details"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <PlannedTransactionsPage
                emptyIcon="cozy"
              />
            </ScrollToTopOnMountWrapper>
          }
          path="future"
        />
        <Route
          element={<SetFilterAndRedirect />}
          path=":accountOrGroupId/:page"
        />
      </Route>
      <Route
        element={
          <Navigate
            replace={true}
            to="../analysis/categories"
          />
        }
        path="categories/*"
      />
      <Route
        element={
          <Navigate
            replace={true}
            to="../analysis/recurrence"
          />
        }
        path="recurrence/*"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <AnalysisPage />
          </ScrollToTopOnMountWrapper>
        }
        path="analysis"
      >
        <Route
          path="categories"
        >
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            index={true}
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            path=":categoryName"
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            path=":categoryName/:subcategoryName"
          />
        </Route>
        <Route
          path="recurrence"
        >
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <withError(RecurrencesPage) />
              </ScrollToTopOnMountWrapper>
            }
            index={true}
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <RecurrencePage />
              </ScrollToTopOnMountWrapper>
            }
            path=":bundleId"
          />
        </Route>
      </Route>
      <Route
        path="settings"
      >
        <Route
          element={
            <Navigate
              replace={true}
              to="../configuration/import"
            />
          }
          path="import"
        />
        <Route
          element={
            <Navigate
              replace={true}
              to="../export"
            />
          }
          path="configuration/export"
        />
        <Route
          element={<Export />}
          path="export"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Settings
                delayContent={0}
              />
            </ScrollToTopOnMountWrapper>
          }
        >
          <Route
            element={<Unknown />}
            index={true}
          />
          <Route
            element={<Memo(AccountsSettings) />}
            path="accounts"
          />
          <Route
            element={<withQuery(withQuery(Groups)) />}
            path="groups"
          />
          <Route
            element={<TagsSettings />}
            path="tags"
          />
          <Route
            element={
              <OutletWrapper
                Component={[Function]}
              />
            }
            path="configuration"
          >
            <Route
              element={<Import />}
              path="import"
            />
          </Route>
        </Route>
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(NewGroupSettings) />
            </ScrollToTopOnMountWrapper>
          }
          path="groups/new"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(ExistingGroupSettings) />
            </ScrollToTopOnMountWrapper>
          }
          path="groups/:groupId"
        />
        <Route
          element={
            <Navigate
              replace={true}
              to="../accounts"
            />
          }
          path="accounts/:accountId"
        />
      </Route>
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <TagPage />
          </ScrollToTopOnMountWrapper>
        }
        path="tag/:tagId"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <Unknown />
          </ScrollToTopOnMountWrapper>
        }
        path="transfers"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <SearchPage />
          </ScrollToTopOnMountWrapper>
        }
        path="search"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <SearchPage />
          </ScrollToTopOnMountWrapper>
        }
        path="search/:search"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <DebugRecurrencePage />
          </ScrollToTopOnMountWrapper>
        }
        path="recurrencedebug"
      />
      <Route
        element={
          <Navigate
            replace={true}
            to="balances"
          />
        }
        path="*"
      />
    </Route>
  </Route>
</Routes>
`;

exports[`App route should have renderExtraRoutes 1`] = `
<Routes>
  <Route
    element={<withClient(UserActionRequired) />}
  >
    <Route
      element={<withQuery(App) />}
      path="/"
    >
      <Route
        element={
          <Navigate
            replace={true}
            to="balances"
          />
        }
        index={true}
      />
      <Route
        path="balances"
      >
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <BalanceWithBanksJobs />
            </ScrollToTopOnMountWrapper>
          }
          index={true}
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(Connect(RawBalanceDetailsPage)) />
            </ScrollToTopOnMountWrapper>
          }
          path="details"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <PlannedTransactionsPage
                emptyIcon="cozy"
              />
            </ScrollToTopOnMountWrapper>
          }
          path="future"
        />
        <Route
          element={<SetFilterAndRedirect />}
          path=":accountOrGroupId/:page"
        />
      </Route>
      <Route
        element={
          <Navigate
            replace={true}
            to="../analysis/categories"
          />
        }
        path="categories/*"
      />
      <Route
        element={
          <Navigate
            replace={true}
            to="../analysis/recurrence"
          />
        }
        path="recurrence/*"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <AnalysisPage />
          </ScrollToTopOnMountWrapper>
        }
        path="analysis"
      >
        <Route
          path="categories"
        >
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            index={true}
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            path=":categoryName"
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            path=":categoryName/:subcategoryName"
          />
        </Route>
        <Route
          path="recurrence"
        >
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <withError(RecurrencesPage) />
              </ScrollToTopOnMountWrapper>
            }
            index={true}
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <RecurrencePage />
              </ScrollToTopOnMountWrapper>
            }
            path=":bundleId"
          />
        </Route>
      </Route>
      <Route
        path="settings"
      >
        <Route
          element={
            <Navigate
              replace={true}
              to="../configuration/import"
            />
          }
          path="import"
        />
        <Route
          element={
            <Navigate
              replace={true}
              to="../export"
            />
          }
          path="configuration/export"
        />
        <Route
          element={<Export />}
          path="export"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Settings
                delayContent={0}
              />
            </ScrollToTopOnMountWrapper>
          }
        >
          <Route
            element={<Unknown />}
            index={true}
          />
          <Route
            element={<Memo(AccountsSettings) />}
            path="accounts"
          />
          <Route
            element={<withQuery(withQuery(Groups)) />}
            path="groups"
          />
          <Route
            element={<TagsSettings />}
            path="tags"
          />
          <Route
            element={
              <OutletWrapper
                Component={[Function]}
              />
            }
            path="configuration"
          >
            <Route
              element={<Import />}
              path="import"
            />
          </Route>
        </Route>
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(NewGroupSettings) />
            </ScrollToTopOnMountWrapper>
          }
          path="groups/new"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(ExistingGroupSettings) />
            </ScrollToTopOnMountWrapper>
          }
          path="groups/:groupId"
        />
        <Route
          element={
            <Navigate
              replace={true}
              to="../accounts"
            />
          }
          path="accounts/:accountId"
        />
      </Route>
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <TagPage />
          </ScrollToTopOnMountWrapper>
        }
        path="tag/:tagId"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <Unknown />
          </ScrollToTopOnMountWrapper>
        }
        path="transfers"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <SearchPage />
          </ScrollToTopOnMountWrapper>
        }
        path="search"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <SearchPage />
          </ScrollToTopOnMountWrapper>
        }
        path="search/:search"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <DebugRecurrencePage />
          </ScrollToTopOnMountWrapper>
        }
        path="recurrencedebug"
      />
      <Route
        element={<NewComponent />}
        path="extra-route"
      />
      <Route
        element={
          <Navigate
            replace={true}
            to="balances"
          />
        }
        path="*"
      />
    </Route>
  </Route>
</Routes>
`;

exports[`App route should have renderExtraRoutesOnly if condition is true 1`] = `
<Routes>
  <Route
    element={<withClient(UserActionRequired) />}
  >
    <Route
      element={<NewComponent />}
      path="extra-route-only"
    />
  </Route>
</Routes>
`;

exports[`App route should not have renderExtraRoutesOnly if condition is false 1`] = `
<Routes>
  <Route
    element={<withClient(UserActionRequired) />}
  >
    <Route
      element={<withQuery(App) />}
      path="/"
    >
      <Route
        element={
          <Navigate
            replace={true}
            to="balances"
          />
        }
        index={true}
      />
      <Route
        path="balances"
      >
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <BalanceWithBanksJobs />
            </ScrollToTopOnMountWrapper>
          }
          index={true}
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(Connect(RawBalanceDetailsPage)) />
            </ScrollToTopOnMountWrapper>
          }
          path="details"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <PlannedTransactionsPage
                emptyIcon="cozy"
              />
            </ScrollToTopOnMountWrapper>
          }
          path="future"
        />
        <Route
          element={<SetFilterAndRedirect />}
          path=":accountOrGroupId/:page"
        />
      </Route>
      <Route
        element={
          <Navigate
            replace={true}
            to="../analysis/categories"
          />
        }
        path="categories/*"
      />
      <Route
        element={
          <Navigate
            replace={true}
            to="../analysis/recurrence"
          />
        }
        path="recurrence/*"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <AnalysisPage />
          </ScrollToTopOnMountWrapper>
        }
        path="analysis"
      >
        <Route
          path="categories"
        >
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            index={true}
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            path=":categoryName"
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <Unknown />
              </ScrollToTopOnMountWrapper>
            }
            path=":categoryName/:subcategoryName"
          />
        </Route>
        <Route
          path="recurrence"
        >
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <withError(RecurrencesPage) />
              </ScrollToTopOnMountWrapper>
            }
            index={true}
          />
          <Route
            element={
              <ScrollToTopOnMountWrapper>
                <RecurrencePage />
              </ScrollToTopOnMountWrapper>
            }
            path=":bundleId"
          />
        </Route>
      </Route>
      <Route
        path="settings"
      >
        <Route
          element={
            <Navigate
              replace={true}
              to="../configuration/import"
            />
          }
          path="import"
        />
        <Route
          element={
            <Navigate
              replace={true}
              to="../export"
            />
          }
          path="configuration/export"
        />
        <Route
          element={<Export />}
          path="export"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Settings
                delayContent={0}
              />
            </ScrollToTopOnMountWrapper>
          }
        >
          <Route
            element={<Unknown />}
            index={true}
          />
          <Route
            element={<Memo(AccountsSettings) />}
            path="accounts"
          />
          <Route
            element={<withQuery(withQuery(Groups)) />}
            path="groups"
          />
          <Route
            element={<TagsSettings />}
            path="tags"
          />
          <Route
            element={
              <OutletWrapper
                Component={[Function]}
              />
            }
            path="configuration"
          >
            <Route
              element={<Import />}
              path="import"
            />
          </Route>
        </Route>
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(NewGroupSettings) />
            </ScrollToTopOnMountWrapper>
          }
          path="groups/new"
        />
        <Route
          element={
            <ScrollToTopOnMountWrapper>
              <Memo(ExistingGroupSettings) />
            </ScrollToTopOnMountWrapper>
          }
          path="groups/:groupId"
        />
        <Route
          element={
            <Navigate
              replace={true}
              to="../accounts"
            />
          }
          path="accounts/:accountId"
        />
      </Route>
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <TagPage />
          </ScrollToTopOnMountWrapper>
        }
        path="tag/:tagId"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <Unknown />
          </ScrollToTopOnMountWrapper>
        }
        path="transfers"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <SearchPage />
          </ScrollToTopOnMountWrapper>
        }
        path="search"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <SearchPage />
          </ScrollToTopOnMountWrapper>
        }
        path="search/:search"
      />
      <Route
        element={
          <ScrollToTopOnMountWrapper>
            <DebugRecurrencePage />
          </ScrollToTopOnMountWrapper>
        }
        path="recurrencedebug"
      />
      <Route
        element={
          <Navigate
            replace={true}
            to="balances"
          />
        }
        path="*"
      />
    </Route>
  </Route>
</Routes>
`;
